Zephyr usbc subsystem

sambhurst opened the pull request:
These pull requests are part of a larger effort to port the Google ChromeOS USB-C Stack to Zephyr. See issue https://github.com/zephyrproject-rtos/zephyr/issues/38371.

Currently the State Machine Framework https://github.com/zephyrproject-rtos/zephyr/pull/38088 and Type-C Port Controller Driver https://github.com/zephyrproject-rtos/zephyr/pull/41856 has merged.
This PR implements the USBC-C Subsystem from which Type-C / Power delivery application can be created. Once this PR merges, Power Delivery Sink only applications can be created.

sambhurst committed:
usb-c: Add Type-C / Power Delivery macros

These macros are needed to implement a USB-C Sink device

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Add USB-C Subsystem with Sink PD Support

This USB-C Subsystem enables an application to include
USB-C Power Delivery Sink functionality.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
samples: usb-c: sink: Add USB-C Subsystem Sink Sample

Implementing USB-C Sink functionality can be difficult.
This sample application serves as an example of how
to create an application with Power Delivery Sink
functionality.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
doc: Document the USB-C Subsystem

Documents the USB-C Subsystem API.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
samples: boards: Remove board specific USB-C Sink sample

This CL removes the board specific USB-C Sink sample in
favor of the generic USB-C Sink sample.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

jfischer-no reviewed (commented):
You need to update includes, e.g. `#include <kernel.h>` -> `#include <zephyr/kernel.h>`
Please consider to place USBC subsystem in the `subsys/usb/usbc`, `subsys/usb` is no longer completely occupied by current device stack, which was moved to `subsys/usb/device`. What will be added else soon is `subsys/usb/device_next `and `subsys/usb/host` (see https://github.com/zephyrproject-rtos/zephyr/issues/42066)

erwango commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -1374,6 +1377,37 @@ static int ucpd_init(const struct device *dev)
 	uint32_t cfg1;
 	int ret;
 
+#ifdef CONFIG_SOC_SERIES_STM32G0X
+	/* Configure CC pins as analog */
+	if ((uint32_t)config->ucpd_port == UCPD1_BASE) {
+		/* UCPD1_CC1 */
+		LL_GPIO_SetPinMode(GPIOA, LL_GPIO_PIN_8, LL_GPIO_MODE_ANALOG);
+		/* UCPD1_CC2 */
+		LL_GPIO_SetPinMode(GPIOB, LL_GPIO_PIN_15, LL_GPIO_MODE_ANALOG);
+	} else {
+		/* UCPD2_CC1 */
+		LL_GPIO_SetPinMode(GPIOD, LL_GPIO_PIN_0, LL_GPIO_MODE_ANALOG);
+		/* UCPD2_CC2 */
+		LL_GPIO_SetPinMode(GPIOD, LL_GPIO_PIN_2, LL_GPIO_MODE_ANALOG);
+	}
+#else
+	/* Disable Dead Dattery */
+	LL_PWR_DisableUCPDDeadBattery();
+#if CONFIG_SOC_SERIES_STM32G4X
+	/* Configure CC pins as analog */
+	/* UCPD1_CC1 */
+	LL_GPIO_SetPinMode(GPIOB, LL_GPIO_PIN_6, LL_GPIO_MODE_ANALOG);
+	/* UCPD1_CC2 */
+	LL_GPIO_SetPinMode(GPIOB, LL_GPIO_PIN_4, LL_GPIO_MODE_ANALOG);
+#else /* CONFIG_SOC_SERIES_STM32U5X || CONFIG_SOC_SERIES_STM32L5X */
+	/* Configure CC pins as analog */
+	/* UCPD1_CC1 */
+	LL_GPIO_SetPinMode(GPIOA, LL_GPIO_PIN_15, LL_GPIO_MODE_ANALOG);
+	/* UCPD1_CC2 */
+	LL_GPIO_SetPinMode(GPIOB, LL_GPIO_PIN_15, LL_GPIO_MODE_ANALOG);
+#endif /* CONFIG_SOC_SERIES_STM32G4X */
+#endif /* CONFIG_SOC_SERIES_STM32G0X */
```
This doesn't scale as I expect PINs in use will vary across boards.
Please use dedicated pinctl API to do this.

sambhurst replied:
I've made the change and it's definitely much nicer.

Setting the CC pins as follows isn't so obvious: pinctrl-0 = <&analog_pb6 &analog_pb4>;
Could you add something like ucpd_cc1_pb6 and ucpd_cc2_pb4 to the modules/hal tx-pinctrl.dtsi?

erwango replied:
> Setting the CC pins as follows isn't so obvious: pinctrl-0 = <&analog_pb6 &analog_pb4>;
Could you add something like ucpd_cc1_pb6 and ucpd_cc2_pb4 to the modules/hal tx-pinctrl.dtsi?

Sure, will do.

erwango replied:
@sambhurst I've pushed https://github.com/zephyrproject-rtos/hal_stm32/pull/144.
To use it, you need to add a commit which will modify west.yml in the following way
```
    - name: hal_stm32
-      revision: 6253094a4147cbd30d2b466c975eb4013ac6be90
+      revision: pull/144/head
      path: modules/hal/stm32
      groups:
        - hal
```

erwango commented on samples/subsys/usbc/sink/README.rst:
```diff
@@ -0,0 +1,108 @@
+Building and Running
+********************
+
+Build and flash as follows, changing ``b_g474e_dpow1_board`` for your board:
```
```suggestion
Build and flash as follows, changing ``b_g474e_dpow1`` for your board:
```

sambhurst replied:
Done

erwango commented on samples/subsys/usbc/sink/README.rst:
```diff
@@ -0,0 +1,108 @@
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/usbc/sink
+   :board: b_g474e_dpow1_board
```
```suggestion
   :board: b_g474e_dpow1
```

sambhurst replied:
Done

sambhurst committed:
dts: bindings: Add pin control to STM32 UCPD DTS bindings file

Add pin control properties to the STM32 UCPD bindings file so that
the pins can be configured in the device tree.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
boards: arm: Update STM32 UCPD nodes with pin control

Assign the pins for the STM32G071b and STM32G081b STM32
UCPD peripherals so the UCPD driver can properly
configure the device.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

erwango commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -1371,9 +1377,17 @@ static int ucpd_init(const struct device *dev)
 {
 	const struct tcpc_config *const config = dev->config;
 	struct tcpc_data *data = dev->data;
+	HAL_StatusTypeDef status;
```
```suggestion
	int status;
```

GeorgeCGV replied:
Better to reuse existing `int ret;`

sambhurst replied:
Done

erwango commented on samples/boards/stm32/usbc/sink/sample.yaml:
```diff
@@ -1,7 +1,7 @@
 sample:
-  name: USBC SINK
+  name: BOARD STM32 USBC SINK
 tests:
-  sample.usbc.sink:
+  sample.board.stm32.usbc.sink:
```
Now that you introduced a generic USBC Sink sample, should we keep this sample ?
If yes, would you mind document in what these samples differ ?

sambhurst replied:
I think the board specific sample should be removed.

sambhurst commented on doc/services/usbc/ucds.rst:
```diff
@@ -0,0 +1,103 @@
+node:
+
+.. literalinclude:: ../../../samples/subsys/usbc/sink/boards/b_g474e_dpow1.overlay
+   :language: guess
```
Done

sambhurst commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,97 @@
+	.resolution = ADC_RESOLUTION,
+};
+
+/* usbc.rst board vbus measurment start */
```
I overlooked it but it's been added

sambhurst commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -14,11 +14,16 @@ LOG_MODULE_REGISTER(ucpd_stm32, CONFIG_USBC_LOG_LEVEL);
 #include <stddef.h>
-#include <zephyr/math/ilog2.h>
-#include <stm32g0xx_ll_system.h>
+#include <math/ilog2.h>
```
Done

erwango reviewed (changes requested):
Few more comments.

One point on sample, but that's nice to have at this point:
It would be great to get 2 boards working with the generic sample.
Then if this involves having variations in the sample, make use of a sample specific Kconfig.

sambhurst commented on samples/subsys/usbc/sink/boards/b_g474e_dpow1.overlay:
```diff
@@ -0,0 +1,21 @@
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+/* usbc.rst usbc-port start */
```
Done

sambhurst replied:
Sorry, I thought I indented these. It's been done.

sambhurst commented on boards/arm/b_g474e_dpow1/b_g474e_dpow1.dts:
```diff
@@ -136,10 +149,6 @@
+	pinctrl-0 = <&ucpd1_cc1_pb6 &ucpd1_cc2_pb4>;
 	pinctrl-names = "default";
-	status = "okay";
 };
```
This was a mistake. I've added it back.

erwango commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,382 @@
+		       pdo.usb_comms_capable);
+		printk("\tUnconstrained Pwr: %d\n",
+		       pdo.unconstrained_power);
+		printk("\tUSB Susspend:      %d\n",
```
```suggestion
		printk("\tUSB Suspend:      %d\n",
```

sambhurst replied:
Done

erwango commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,95 @@
+#if DT_HAS_COMPAT_STATUS_OKAY(usbc_port)
+
+#define VBUS DT_PATH(vbus)
+#define LED_LABEL       DT_GPIO_LABEL(LED0_NODE, gpios)
+#define LED_PIN         DT_GPIO_PIN(LED0_NODE, gpios)
+#define LED_FLAGS       DT_GPIO_FLAGS(LED0_NODE, gpios)
```
Not required anymore ?

sambhurst replied:
Removed

mbolivar-nordic commented on dts/bindings/tcpc/st,stm32-ucpd.yaml:
```diff
@@ -69,3 +69,9 @@ properties:
         Valid range: 1 - 64
+
+    pinctrl-0:
+      required: true
```
This commit order breaks git bisect.

sambhurst replied:
I changed the commit order

mbolivar-nordic commented on boards/arm/stm32g071b_disco/stm32g071b_disco.dts:
```diff
@@ -177,6 +177,8 @@
 	 */
 	psc-ucpdclk = <1>;
 	hbitclkdiv = <27>;
+	pinctrl-0 = <&ucpd1_cc1_pa8 &ucpd1_cc2_pb15>;
```
... since the property is made required before it is defined in these boards.

sambhurst replied:
I changed the commit order

sambhurst replied:
Sorry, I didn't fully understand what you meant.

The following PR must merge before I can use the properties: https://github.com/zephyrproject-rtos/hal_stm32/pull/144/commits

mbolivar-nordic commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -15,10 +15,15 @@ LOG_MODULE_REGISTER(ucpd_stm32, CONFIG_USBC_LOG_LEVEL);
 
 #include "ucpd_stm32_priv.h"
 
+PINCTRL_DT_INST_DEFINE(0);
+static const struct pinctrl_dev_config *usbc_pcfg =
+					PINCTRL_DT_INST_DEV_CONFIG_GET(0);
```
This looks a little strange - @gmarull could you please sanity check this?

gmarull replied:
driver seems to be multi-instance, however, this breaks multi-instance support. Please move this to `TCPC_DRIVER_INIT` and store a reference to `drv_config_##inst`

sambhurst replied:
Done

mbolivar-nordic commented on dts/bindings/usbc/usbc-port.yaml:
```diff
@@ -0,0 +1,16 @@
+# Copyright 2022 The Chromium OS Authors
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Common fields for USBC ports
+
+compatible: "usbc-port"
```
I'm confused. It looks like this is meant to be an include file, not actually a compatible that is used within "real" devicetree files. Is that right? If so, this compatible should be removed.

sambhurst replied:
I updated the commit. This was not meant to be a common file

mbolivar-nordic commented on subsys/usb/usbc/Kconfig:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Workaround for not being able to have commas in macro arguments
+DT_COMPAT_USBC_PORT := usbc-port
+
+menuconfig USBC_STACK
+	bool "USBC Stack Support"
+	default $(dt_compat_enabled,$(DT_COMPAT_USBC_PORT))
```
OK, now I see this is being used like a "real" compatible, but then the comment in the binding's description that it's "common fields" doesn't make sense. Something is wrong here. I would have expected individual driver IPs to have their own compatibles in their own bindings. Each of those could `select` the right Kconfig symbol to turn on the common infrastructure in addition to doing anything IP-specific in their own drive rfiles. Unless there is some sort of common IP for USB type C, the way 16550 is for serial?

sambhurst replied:
I updated the comment in usbc-port.yaml

mbolivar-nordic commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,95 @@
+{
+	dev_adc = DEVICE_DT_GET(DT_IO_CHANNELS_CTLR(VBUS));
+	if (!device_is_ready(dev_adc)) {
+		printk("ADC device not found\n");
```
Seems like logging should be used instead, no? A general comment about this PR.

sambhurst replied:
Updated to use LOG_INF

mbolivar-nordic commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,382 @@
+
+	/* usbc.rst config board start */
+	/* Config the board */
+	ret = board_config();
```
No way, let's not go down the road of board specific C files called into by main like this. This is the sort of thing that should happen before main should even run, using a SYS_INIT or similar defined by the driver or subsystem layer.

sambhurst replied:
I agree and this has been updated

mbolivar-nordic reviewed (changes requested):
Hey, finally got a chance to give this a quick skim now that v3.1 is out. Here are some initial comments.

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,95 @@
+#define LED_FLAGS       DT_GPIO_FLAGS(LED0_NODE, gpios)
+
+/* Common settings supported by most ADCs */
+#define ADC_RESOLUTION          12
+#define ADC_GAIN                ADC_GAIN_1
+#define ADC_REFERENCE           ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME    ADC_ACQ_TIME_DEFAULT
+#define ADC_REF_MV              3300
```
you can now use the recently added adc dt-spec facilities to make code more portable
https://github.com/zephyrproject-rtos/zephyr/pull/43030

and soon:
https://github.com/zephyrproject-rtos/zephyr/pull/46110

sambhurst replied:
Updated

gmarull commented on subsys/usb/usbc/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+zephyr_library_sources_ifdef(CONFIG_USBC_STACK usbc_timer.c usbc_stack.c usbc_tc.c usbc_prl.c usbc_pe.c)
+
+if(CONFIG_USBC_STACK)
+  zephyr_include_directories(${ZEPHYR_BASE}/subsys/usbc)
```
why can't headers be placed under `include/` if they are public?

sambhurst replied:
I've removed the include part

gmarull commented on subsys/usb/usbc/usbc_pe.h:
```diff
@@ -0,0 +1,217 @@
+
+#include <zephyr/kernel.h>
+#include <usbc/usbc.h>
+#include <drivers/usbc/usbc_pd.h>
+#include <drivers/usbc/usbc_tc.h>
+#include <smf.h>
```
please prefix with `zephyr/`

sambhurst replied:
done

gmarull commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,299 @@
+			if (request == PRIV_PORT_REQUEST_SUSPEND) {			     \
+				k_thread_suspend(port->port_thread);			     \
+			}								     \
+			k_msleep(5);							     \
```
magic sleep here?

sambhurst replied:
I changed the magic number to a Kconfig option

gmarull commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,299 @@
+static const struct usbc_subsystem_api usbc_api;
+static int usbc_subsys_init(const struct device *dev);
+
+#define THREAD_STACK(inst) K_THREAD_STACK_DEFINE(my_stack_area_##inst, CONFIG_USBC_STACK_SIZE);
```
better avoid `;`, so that callers can do `THREAD_STACK(...);`

erwango commented on boards/arm/b_g474e_dpow1/b_g474e_dpow1.dts:
```diff
@@ -136,6 +162,8 @@
 	 */
 	psc-ucpdclk = <1>;
 	hbitclkdiv = <27>;
+	pinctrl-0 = <&ucpd1_cc1_pb6 &ucpd1_cc2_pb4>;
```
It seems the west.yml commit which adds usbc pins support is gone.
Please put it back (prior to this commit) and since it is now merged , you can refer directly to the SHA1 of the hal_stm32 commit: c87358a3e9b99e0f8f18816f1a74777fe80d1ab6

sambhurst replied:
Done

sambhurst commented on boards/arm/stm32g081b_eval/board.c:
```diff
@@ -0,0 +1,84 @@
+	ret = adc_raw_to_millivolts(adc_channel.vref_mv,
+				    adc_channel.channel_cfg.gain,
+				    adc_channel.resolution,
+				    &value);
```
Done

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,83 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(board, LOG_LEVEL_DBG);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(usbc_port)
```
this can be avoided if this code is part of a sample, and sample provides a proper filter. for example, you can add to `sample.yaml`:

```yaml
filter: dt_compat_enabled("usbc-port")
```

sambhurst replied:
I thought sample.yaml was used during testing. Could you give a more complete example of how "filter" could be used to exclude board.c when the usbc-port is not defined?

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,83 @@
+
+#if DT_HAS_COMPAT_STATUS_OKAY(usbc_port)
+
+#define VBUS DT_PATH(vbus)
```
suggestion: better rely on nodelabel, so there's no restriction on where to place vbus

sambhurst replied:
Done

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,83 @@
+/* Data of ADC io-channels specified in devicetree. */
+static const struct adc_dt_spec adc_channel = ADC_DT_SPEC_GET(VBUS);
+
+static struct adc_sequence sequence = {
+	.channels = BIT(adc_channel.channel_id),
+	.buffer = &sample_buffer,
+	/* buffer size in bytes, not number of samples */
+	.buffer_size = sizeof(sample_buffer),
+	.resolution = adc_channel.resolution,
+};
```
note: there's `adc_sequence_init_dt` to populate certain fields from DT.

sambhurst replied:
Done

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,83 @@
+		return -ENODEV;
+	}
+
+	adc_channel_setup(adc_channel.dev, &adc_channel.channel_cfg);
```
`adc_channel_setup_dt`

sambhurst replied:
Done

gmarull commented on boards/arm/b_g474e_dpow1/board.c:
```diff
@@ -0,0 +1,83 @@
+/*
```
this code seems to belong to a sample, not board level.

sambhurst replied:
This code was placed at the board level because the ADC reading of VBUS and USB-C configuration could very between devices. Although for these two samples, the code is similar.

gmarull commented on boards/arm/stm32g081b_eval/board.c:
```diff
@@ -0,0 +1,81 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(board, LOG_LEVEL_DBG);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(usbc_port)
+
+#define VBUS DT_PATH(vbus)
+
+/* Common settings supported by most ADCs */
+
+static int16_t sample_buffer;
+
+static const uint32_t output_ohm = DT_PROP(VBUS, output_ohms);
+static const uint32_t full_ohm = DT_PROP(VBUS, full_ohms);
+
+/* Data of ADC io-channels specified in devicetree. */
```
this code seems now to be a copy from the other board, should likely be in the sample (and sample can be made non-board specific)

sambhurst replied:
Yes, for these two boards, the ADC code is similar but other boards could read VBUS very differently.

gmarull replied:
why not create a sample level module for boards that use ADC? it can be compiled based on a sample level option choice, e.g. `MYSAMPLE_MEASURE_ADC`, which can in turn be enabled by board .conf overlay.

sambhurst replied:
I agree with you and refactored the code. I refactored the VBUS measurement and control into a device driver.

sambhurst committed:
usb-c: Add USB-C VBUS driver

USB-C VBUS can be measured and controlled by different devices.
For example, a TypeC Port Controller (TCPC) or an ADC and GPIO
can be used. This driver is used to abstracts the underlying
hardware used for USB VBUS measurement and control.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Add USB-C VBUS ADC driver

This commit adds a VBUS driver the uses an ADC connected
to a voltage divider to measure VBUS.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Remove VBUS measurement and control

A USB-C VBUS driver will be used to measure VBUS
instead of accessing the measurement hardware
directly.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

carlescufi commented:
@jfischer-no could you please review?

gmarull commented on include/zephyr/drivers/usbc/usbc_vbus.h:
```diff
@@ -0,0 +1,128 @@
+ * @param dev        Runtime device structure
+ * @param meas       pointer where the measured VBUS voltage is stored
+ *
+ * @return 0 on success
+ * @return -EIO on failure
```
please use `@retval` (applies to other cases as well)

sambhurst replied:
Done

gmarull commented on include/zephyr/drivers/usbc/usbc_vbus.h:
```diff
@@ -0,0 +1,128 @@
+extern "C" {
+#endif
+
+__subsystem struct vbus_driver_api {
```
I think `__subsystem` is only needed for userspace, doesn't seem this API supports it.

sambhurst replied:
Done

gmarull commented on dts/bindings/usbc/vbus-voltage-divider-adc.yaml:
```diff
@@ -0,0 +1,13 @@
+include: [base.yaml, voltage-divider.yaml]
+
+properties:
+    discharge-gpios:
+       type: phandle-array
+       required: false
```
where is this used? can't see any usage in https://github.com/zephyrproject-rtos/zephyr/commit/70f420f53a8d92c0ce80cb6ecaef5c8aba966632

sambhurst replied:
It's used here [b5dfa79](https://github.com/zephyrproject-rtos/zephyr/pull/45601/commits/b5dfa799b88dc87965ccb79faba655710688cc1c)

		/* Pin B14 is used to control VBUS Discharge for Port2 */
		discharge-gpios = <&gpiob 14 GPIO_ACTIVE_HIGH>;

gmarull commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,390 @@
+	WAIT_VCONN_SWAP,
+};
+
+typedef int (*policy_cb_get_snk_cap_t)(const struct device *dev, uint32_t **pdos, int *num_pdos);
+typedef void (*policy_cb_set_src_cap_t)(const struct device *dev, uint32_t *pdos, int num_pdos);
+typedef bool (*policy_cb_check_t)(const struct device *dev, enum policy_check_t policy_check);
+typedef bool (*policy_cb_wait_notify_t)(const struct device *dev, enum policy_wait_t wait_notify);
+typedef void (*policy_cb_notify_t)(const struct device *dev, enum policy_notify_t policy_notify);
+typedef uint32_t (*policy_cb_get_request_data_object_t)(const struct device *dev);
+typedef uint32_t (*policy_cb_get_ps_transition_time_t)(const struct device *dev);
+typedef bool (*policy_cb_is_snk_at_default_level_t)(const struct device *dev);
+
+__subsystem struct usbc_subsystem_api {
+	int (*start)(const struct device *dev);
+	int (*suspend)(const struct device *dev);
+	int (*request)(const struct device *dev, enum policy_request_t req);
+	void (*set_dpm_data)(const struct device *dev, void *dpm_data);
+	void * (*get_dpm_data)(const struct device *dev);
+	void (*set_vconn_control_cb)(const struct device *dev,
+				     tcpc_vconn_control_cb_t set_vconn_cb);
+	void (*set_policy_cb_get_snk_cap)(const struct device *dev,
+					  policy_cb_get_snk_cap_t policy_cb_get_snk_cap);
+	void (*set_policy_cb_set_src_cap)(const struct device *dev,
+					  policy_cb_set_src_cap_t policy_cb_set_src_cap);
+	void (*set_policy_cb_check)(const struct device *dev, policy_cb_check_t policy_cb_check);
+	void (*set_policy_cb_wait_notify)(const struct device *dev,
+					  policy_cb_wait_notify_t policy_cb_wait_notify);
+	void (*set_policy_cb_notify)(const struct device *dev,
+				     policy_cb_notify_t policy_cb_notify);
+	void (*set_policy_cb_get_request_data_object)(const struct device *dev,
+			policy_cb_get_request_data_object_t policy_cb_get_request_data_object);
+	void (*set_policy_cb_get_ps_transition_time)(const struct device *dev,
+			policy_cb_get_ps_transition_time_t policy_cb_get_ps_transition_time);
+	void (*set_policy_cb_is_snk_at_default_level)(const struct device *dev,
+			policy_cb_is_snk_at_default_level_t callback);
+};
```
if internal, enclose under `/** @cond INTERNAL_HIDDEN */` (see other headers as reference, e.g. pwm.h)

sambhurst replied:
Done

gmarull reviewed (commented):
a few minor nitpicks

sambhurst committed:
usb-c: tcpc: Changes the @return tag to @retval

The "@return" command should be used for the arbitrary return
description, and "@retval <return value>" for the description
of named return value

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented on drivers/usbc/vbus/vbus_voltage_divider_adc.c:
```diff
@@ -0,0 +1,142 @@
+
+	if (meas == NULL) {
+		return -EFAULT;
+	}
```
Done

sambhurst commented on drivers/usbc/vbus/vbus_voltage_divider_adc.c:
```diff
@@ -0,0 +1,142 @@
+	ret = adc_raw_to_millivolts_dt(&config->adc_channel, &value);
+	if (ret != 0) {
+		LOG_INF("Scaling ADC failed with error %d.", ret);
+		return -EIO;
```
Done

sambhurst commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,390 @@
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/usbc/usbc_tcpc.h>
+#include <drivers/usbc/usbc_vbus.h>
```
Done

sambhurst commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,390 @@
+typedef uint32_t (*policy_cb_get_ps_transition_time_t)(const struct device *dev);
+typedef bool (*policy_cb_is_snk_at_default_level_t)(const struct device *dev);
+
+__subsystem struct usbc_subsystem_api {
```
Done

sambhurst commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,271 @@
+#include "usbc_stack.h"
+
+#define LOG_MODULE_NAME usbc_stack
+LOG_MODULE_REGISTER(usbc_stack);
```
I removed it

sambhurst commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,271 @@
+	/* Make sure TCPC is ready */
+	if (!device_is_ready(tcpc)) {
+		LOG_ERR("TCPC NOT READY");
+		return -EIO;
```
Done

sambhurst commented on subsys/usb/usbc/usbc_timer.c:
```diff
@@ -0,0 +1,75 @@
+/**
+ * @brief Starts a timer
+ */
+void usbc_timer_start(struct usbc_timer_t *usbc_timer)
```
I removed them

jfischer-no commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,270 @@
+		port->port_thread = k_thread_create(&port->thread_data,			\
+				my_stack_area_##inst,					\
+				K_THREAD_STACK_SIZEOF(my_stack_area_##inst),		\
+				run_usbc_##inst,					\
+				(void *)dev, 0, 0,					\
+				CONFIG_USBC_THREAD_PRIORITY, K_ESSENTIAL, K_NO_WAIT);	\
+		k_thread_suspend(port->port_thread);					\
+	}
+
+#define USBC_SUBSYS_INIT(inst)						 \
+	THREAD_STACK(inst);						 \
+	PORT_THREAD(inst)						 \
+	CREATE_THREAD(inst)						 \
+	static struct tc_sm_t tc_##inst;				 \
+	static struct policy_engine pe_##inst;				 \
+	static struct protocol_layer_rx_t prl_rx_##inst;		 \
+	static struct protocol_layer_tx_t prl_tx_##inst;		 \
+	static struct protocol_hard_reset_t prl_hr_##inst;		 \
+	static struct usbc_port_data usbc_port_data_##inst = {		 \
+		.tc = &tc_##inst,					 \
+		.pe = &pe_##inst,					 \
+		.prl_rx = &prl_rx_##inst,				 \
+		.prl_tx = &prl_tx_##inst,				 \
+		.prl_hr = &prl_hr_##inst,				 \
+		.tcpc = DEVICE_DT_GET(DT_INST_PROP(inst, tcpc)),	 \
+		.vbus = DEVICE_DT_GET(DT_INST_PROP(inst, vbus)),	 \
+	};								 \
+	static const struct usbc_port_config usbc_port_config_##inst = { \
+		.create_thread = create_thread_##inst,			 \
+	};								 \
+	DEVICE_DT_INST_DEFINE(inst,					 \
+			      &usbc_subsys_init,			 \
+			      NULL,					 \
+			      &usbc_port_data_##inst,			 \
+			      &usbc_port_config_##inst,			 \
+			      APPLICATION,				 \
+			      CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,	 \
+			      &usbc_api);
+
+DT_INST_FOREACH_STATUS_OKAY(USBC_SUBSYS_INIT)
```
This can be much clearer, please apply & squash https://github.com/jfischer-no/zephyr/commit/c24adf6f112bd45e7ddd3b9f990787a925ed61df.patch

```diff
diff --git a/subsys/usb/usbc/usbc_stack.c b/subsys/usb/usbc/usbc_stack.c
index 7890a2b1712..eed846b6e54 100644
--- a/subsys/usb/usbc/usbc_stack.c
+++ b/subsys/usb/usbc/usbc_stack.c
@@ -20,68 +20,81 @@ LOG_MODULE_REGISTER(usbc_stack);
 static const struct usbc_subsystem_api usbc_api;
 static int usbc_subsys_init(const struct device *dev);
 
-#define THREAD_STACK(inst) K_THREAD_STACK_DEFINE(my_stack_area_##inst, CONFIG_USBC_STACK_SIZE)
-
-#define PORT_THREAD(inst) void run_usbc_##inst(void *port_dev, void *unused1, void *unused2) \
-       {                                                                                    \
-               const struct device *dev = (const struct device *)port_dev;                  \
-               struct usbc_port_data *port = dev->data;                                     \
-               struct request_value *req;                                                   \
-               int32_t request;                                                             \
-                                                                                            \
-               while (1) {                                                                  \
-                       req = k_fifo_get(&port->request_fifo, K_NO_WAIT);                    \
-                       request = (req != NULL) ? req->val : REQUEST_NOP;                    \
-                       pe_run(dev, request);                                        \
-                       prl_run(dev);                                                \
-                       tc_run(dev, request);                                        \
-                       if (request == PRIV_PORT_REQUEST_SUSPEND) {                          \
-                               k_thread_suspend(port->port_thread);                         \
-                       }                                                                    \
-                       k_msleep(CONFIG_USBC_STATE_MACHINE_CYCLE_TIME);                      \
-               }                                                                            \
-       }
-#define CREATE_THREAD(inst) void create_thread_##inst(const struct device *dev)                \
-       {                                                                               \
-               struct usbc_port_data *port = dev->data;                                \
-                                                                                       \
-               port->port_thread = k_thread_create(&port->thread_data,                 \
-                               my_stack_area_##inst,                                   \
-                               K_THREAD_STACK_SIZEOF(my_stack_area_##inst),            \
-                               run_usbc_##inst,                                        \
-                               (void *)dev, 0, 0,                                      \
-                               CONFIG_USBC_THREAD_PRIORITY, K_ESSENTIAL, K_NO_WAIT);   \
-               k_thread_suspend(port->port_thread);                                    \
+static ALWAYS_INLINE void usbc_handler(void *port_dev)
+{
+       const struct device *dev = (const struct device *)port_dev;
+       struct usbc_port_data *port = dev->data;
+       struct request_value *req;
+       int32_t request;
+
+       req = k_fifo_get(&port->request_fifo, K_NO_WAIT);
+       request = (req != NULL) ? req->val : REQUEST_NOP;
+       pe_run(dev, request);
+       prl_run(dev);
+       tc_run(dev, request);
+
+       if (request == PRIV_PORT_REQUEST_SUSPEND) {
+               k_thread_suspend(port->port_thread);
        }
 
-#define USBC_SUBSYS_INIT(inst)                                          \
-       THREAD_STACK(inst);                                              \
-       PORT_THREAD(inst)                                                \
-       CREATE_THREAD(inst)                                              \
-       static struct tc_sm_t tc_##inst;                                 \
-       static struct policy_engine pe_##inst;                           \
-       static struct protocol_layer_rx_t prl_rx_##inst;                 \
-       static struct protocol_layer_tx_t prl_tx_##inst;                 \
-       static struct protocol_hard_reset_t prl_hr_##inst;               \
-       static struct usbc_port_data usbc_port_data_##inst = {           \
-               .tc = &tc_##inst,                                        \
-               .pe = &pe_##inst,                                        \
-               .prl_rx = &prl_rx_##inst,                                \
-               .prl_tx = &prl_tx_##inst,                                \
-               .prl_hr = &prl_hr_##inst,                                \
-               .tcpc = DEVICE_DT_GET(DT_INST_PROP(inst, tcpc)),         \
-               .vbus = DEVICE_DT_GET(DT_INST_PROP(inst, vbus)),         \
-       };                                                               \
-       static const struct usbc_port_config usbc_port_config_##inst = { \
-               .create_thread = create_thread_##inst,                   \
-       };                                                               \
-       DEVICE_DT_INST_DEFINE(inst,                                      \
-                             &usbc_subsys_init,                         \
-                             NULL,                                      \
-                             &usbc_port_data_##inst,                    \
-                             &usbc_port_config_##inst,                  \
-                             APPLICATION,                               \
-                             CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,       \
+       k_msleep(CONFIG_USBC_STATE_MACHINE_CYCLE_TIME);
+}
+
+#define USBC_SUBSYS_INIT(inst)                                                 \
+       K_THREAD_STACK_DEFINE(my_stack_area_##inst,                             \
+                             CONFIG_USBC_STACK_SIZE);                          \
+                                                                               \
+       static struct tc_sm_t tc_##inst;                                        \
+       static struct policy_engine pe_##inst;                                  \
+       static struct protocol_layer_rx_t prl_rx_##inst;                        \
+       static struct protocol_layer_tx_t prl_tx_##inst;                        \
+       static struct protocol_hard_reset_t prl_hr_##inst;                      \
+                                                                               \
+       static void run_usbc_##inst(void *port_dev,                             \
+                                   void *unused1,                              \
+                                   void *unused2)                              \
+       {                                                                       \
+               while (1) {                                                     \
+                       usbc_handler(port_dev);                                 \
+               }                                                               \
+       }                                                                       \
+                                                                               \
+       static void create_thread_##inst(const struct device *dev)              \
+       {                                                                       \
+               struct usbc_port_data *port = dev->data;                        \
+                                                                               \
+               port->port_thread = k_thread_create(&port->thread_data,         \
+                               my_stack_area_##inst,                           \
+                               K_THREAD_STACK_SIZEOF(my_stack_area_##inst),    \
+                               run_usbc_##inst,                                \
+                               (void *)dev, 0, 0,                              \
+                               CONFIG_USBC_THREAD_PRIORITY,                    \
+                               K_ESSENTIAL,                                    \
+                               K_NO_WAIT);                                     \
+               k_thread_suspend(port->port_thread);                            \
+       }                                                                       \
+                                                                               \
+       static struct usbc_port_data usbc_port_data_##inst = {                  \
+               .tc = &tc_##inst,                                               \
+               .pe = &pe_##inst,                                               \
+               .prl_rx = &prl_rx_##inst,                                       \
+               .prl_tx = &prl_tx_##inst,                                       \
+               .prl_hr = &prl_hr_##inst,                                       \
+               .tcpc = DEVICE_DT_GET(DT_INST_PROP(inst, tcpc)),                \
+               .vbus = DEVICE_DT_GET(DT_INST_PROP(inst, vbus)),                \
+       };                                                                      \
+                                                                               \
+       static const struct usbc_port_config usbc_port_config_##inst = {        \
+               .create_thread = create_thread_##inst,                          \
+       };                                                                      \
+                                                                               \
+       DEVICE_DT_INST_DEFINE(inst,                                             \
+                             &usbc_subsys_init,                                \
+                             NULL,                                             \
+                             &usbc_port_data_##inst,                           \
+                             &usbc_port_config_##inst,                         \
+                             APPLICATION,                                      \
+                             CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,              \
                              &usbc_api);
 
 DT_INST_FOREACH_STATUS_OKAY(USBC_SUBSYS_INIT)
```

sambhurst replied:
Done

jfischer-no commented on subsys/usb/usbc/usbc_tc.c:
```diff
@@ -0,0 +1,518 @@
+/**
+ * @brief These macros SET, CLEAR, and CHECK, a Type-C Layer Flag
+ */
+#define SET_FLAG(obj, flag) atomic_set_bit(&obj->flags, flag)
```
^

sambhurst replied:
Done

sambhurst commented on include/zephyr/drivers/usbc/usbc_vbus.h:
```diff
@@ -0,0 +1,128 @@
+#endif
+
+struct vbus_driver_api {
+	bool (*vbus_check_level)(const struct device *dev, enum tc_vbus_level level);
```
Done

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+ */
+void pe_subsys_init(const struct device *dev)
+{
+	struct usbc_port_data *port = dev->data;
```
Done

sambhurst commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,392 @@
+	int (*request)(const struct device *dev, enum policy_request_t req);
+	void (*set_dpm_data)(const struct device *dev, void *dpm_data);
+	void * (*get_dpm_data)(const struct device *dev);
+	void (*set_vconn_control_cb)(const struct device *dev,
```
Removed

sambhurst commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,392 @@
+				     policy_cb_notify_t policy_cb_notify);
+	void (*set_policy_cb_get_request_data_object)(const struct device *dev,
+			policy_cb_get_request_data_object_t policy_cb_get_request_data_object);
+	void (*set_policy_cb_get_ps_transition_time)(const struct device *dev,
```
Removed

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+#include <zephyr/usbc/usbc.h>
+#include <zephyr/drivers/usbc/usbc_pd.h>
+#include <logging/log.h>
+LOG_MODULE_DECLARE(usbc_stack);
```
Done

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+/**
+ * @brief These macros SET, CLEAR, and CHECK, a Policy Engine Layer Flag
+ */
+#define SET_FLAG(obj, flag) atomic_set_bit(&obj->flags, flag)
```
Done

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+/**
+ * @brief Request desired voltage from source.
+ */
+static void pe_send_request_msg(const struct device *dev, uint32_t rdo)
```
Done

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+ * @brief Check if a specific control message was received
+ */
+static bool received_control_message(const struct device *dev,
+				     union pd_header header, enum pd_ctrl_msg_type mt)
```
Done

sambhurst commented on subsys/usb/usbc/usbc_pe.c:
```diff
@@ -0,0 +1,1703 @@
+ * @brief Check if a specific data message was received
+ */
+static bool received_data_message(const struct device *dev,
+				  union pd_header header, enum pd_data_msg_type mt)
```
Done

sambhurst commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,392 @@
+typedef uint32_t (*policy_cb_get_ps_transition_time_t)(const struct device *dev);
+typedef bool (*policy_cb_is_snk_at_default_level_t)(const struct device *dev);
+
+struct usbc_subsystem_api {
```
It's been removed

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+#include <zephyr/usbc/usbc.h>
+#include <zephyr/drivers/usbc/usbc_pd.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(usbc_stack);
```
Done

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+/**
+ * @brief These macros SET, CLEAR, and CHECK, a Protocol Layer Flag
+ */
+#define SET_FLAG(obj, flag) atomic_set_bit(&obj->flags, flag)
```
Done

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+}
+
+/**
+ * @brief Set revision for the give packet type. This function is called from the Policy Engine.
```
Done

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+/**
+ * @brief Alert Handler called by the TCPC driver
+ */
+static void alert_handler(const struct device *tcpc, void *port_dev, enum tcpc_alert alert)
```
The types need to stay as is until the TCPC driver API is updated.

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+	if (msg_type != PD_CTRL_PING) {
+		int p;
+
+		LOG_PRINTK("RECV %04x/%d ", rx_emsg->header.raw_value, num_data_objs);
```
Changed to using LOG_INF

sambhurst commented on subsys/usb/usbc/usbc_prl.c:
```diff
@@ -0,0 +1,1195 @@
+}
+
+/**
+ * @brief This function implements both the Protocol Layer Message Reception State Machine.
```
Done

sambhurst commented on subsys/usb/usbc/usbc_prl.h:
```diff
@@ -0,0 +1,183 @@
+ * @param type SOP* packet sent from port partner
+ * @param rev Revision sent from the port partner
+ */
+void prl_set_rev(const struct device *dev, enum pd_packet_type type, enum pd_rev_type rev);
```
Done

sambhurst commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,270 @@
+
+#include "usbc_stack.h"
+
+LOG_MODULE_REGISTER(usbc_stack);
```
Done

sambhurst commented on subsys/usb/usbc/usbc_stack.c:
```diff
@@ -0,0 +1,270 @@
+
+#define PORT_THREAD(inst) void run_usbc_##inst(void *port_dev, void *unused1, void *unused2) \
+	{										     \
+		const struct device *dev = (const struct device *)port_dev;		     \
+		struct usbc_port_data *port = dev->data;				     \
+		struct request_value *req;						     \
+		int32_t request;							     \
+											     \
+		while (1) {								     \
+			req = k_fifo_get(&port->request_fifo, K_NO_WAIT);		     \
+			request = (req != NULL) ? req->val : REQUEST_NOP;		     \
+			pe_run(dev, request);					     \
+			prl_run(dev);						     \
+			tc_run(dev, request);					     \
+			if (request == PRIV_PORT_REQUEST_SUSPEND) {			     \
+				k_thread_suspend(port->port_thread);			     \
+			}								     \
+			k_msleep(CONFIG_USBC_STATE_MACHINE_CYCLE_TIME);			     \
+		}									     \
```
I initially thought about using a workqueue but decided against it because the USBC timing requirements might not be met if the system is loaded.

As for tolerance, I've only tested the code running with one USBC port.

sambhurst commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,362 @@
+ * @brief Port Policy Callback to Get Sink Capabilities
+ */
+static int usbc_port1_policy_cb_get_snk_cap(const struct device *dev,
+					    uint32_t **pdos, int *num_pdos)
```
Done

sambhurst commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,362 @@
+
+/* usbc.rst callbacks start */
+/**
+ * @brief Port Policy Callback to Get Sink Capabilities
```
Done

sambhurst commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,362 @@
+ */
+static void init_port1_snk_caps(void)
+{
+	union pd_fixed_supply_pdo_sink fixed_supply_pdo_sink;
+
+	fixed_supply_pdo_sink.operational_current = PD_CONVERT_MA_TO_FIXED_PDO_CURRENT(100);
+	fixed_supply_pdo_sink.voltage = PD_CONVERT_MV_TO_FIXED_PDO_VOLTAGE(5000);
+	fixed_supply_pdo_sink.frs_required = FRS_NOT_SUPPORTED;
+	fixed_supply_pdo_sink.dual_role_data = false;
+	fixed_supply_pdo_sink.usb_comms_capable = false;
+	fixed_supply_pdo_sink.unconstrained_power = false;
+	fixed_supply_pdo_sink.higher_capability = false;
+	fixed_supply_pdo_sink.dual_role_power = false;
+	fixed_supply_pdo_sink.type = PDO_FIXED;
+
+	port1_data.snk_caps[0] = fixed_supply_pdo_sink.raw_value;
+	port1_data.snk_cap_cnt = 1;
```
Done

sambhurst commented on samples/subsys/usbc/sink/src/main.c:
```diff
@@ -0,0 +1,362 @@
+/**
+ * @brief Port Policy Callback to get Request Data Object (RDO)
+ */
+static uint32_t usbc_port1_policy_cb_get_request_data_object(const struct device *dev)
```
Done

jfischer-no reviewed (commented):
Thank you, just two comments.

sambhurst commented:
@mbolivar-nordic could you take another look when you have time. Thanks

mbolivar-nordic reviewed (changes requested):
Sorry for the delays, here's a new round of review, again mostly focused on DT.

sambhurst commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -49,7 +49,7 @@ static int ucpd_txorderset[] = {
  * @brief Test for a goodCRC message
  *
- * @returns true if message is goodCRC, else false
+ * @retval true if message is goodCRC, else false
```
Thanks and I've updated the commit message.

mbolivar-nordic replied:
Please note this part:

> This same issue appears throughout the pull request.

It wasn't just this commit message; it was a general note about the commit messages, especially later on in the PR where the "real work" gets done

sambhurst replied:
Understood. Thanks again

sambhurst commented on drivers/usbc/vbus/vbus_adc.c:
```diff
@@ -0,0 +1,138 @@
+	static struct vbus_data drv_data_##inst;			\
+	static const struct vbus_config drv_config_##inst = {		\
+		.output_ohm = DT_INST_PROP(inst, output_ohms),		\
+		.full_ohm = DT_INST_PROP(inst, full_ohms),		\
```
Yes this was an oversight and has been fixed.

sambhurst commented on dts/bindings/usbc/vbus-adc.yaml:
```diff
@@ -0,0 +1,13 @@
+
+compatible: "vbus-adc"
+
+include: [base.yaml, voltage-divider.yaml]
+
+properties:
+    discharge-gpios:
```
I've added the dicharge-gpios and power-gpios to the driver.

mbolivar-nordic commented on doc/services/index.rst:
```diff
@@ -26,6 +26,7 @@ OS Services
    task_wdt/index.rst
    tfm/index
    usb/index.rst
+   usbc/index.rst
```
My guess would have been under doc/hardware/peripherals/ if this is really a device driver API for usb type C peripheals, FWIW.

sambhurst replied:
The USBC subsystem is more than a device driver. It allows you to create applications with USBC Power Delivery functionality. What do you think about placing it under doc/connectivity?

jfischer-no commented on drivers/usbc/vbus/Kconfig.vbus_adc:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2022 The Chromium OS Authors
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_VBUS_ADC := vbus-adc
```
Please take a look at support for "devicetree Kconfig symbol" introduced by https://github.com/zephyrproject-rtos/zephyr/pull/47196

sambhurst replied:
Thanks. I've updated the Kconfig

sambhurst committed:
drivers: usb-c: Update TCPC driver to use devicetree Kconfig symbol

Update TCPC driver to use DT_HAS_<compat>_ENABLED Kconfig symbol
to expose the driver and enable it by default based on devicetree.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@mbolivar-nordic could you take another look when you have time? Thanks

erwango commented on boards/arm/b_g474e_dpow1/b_g474e_dpow1.dts:
```diff
@@ -7,6 +7,7 @@
 /dts-v1/;
 #include <st/g4/stm32g474Xe.dtsi>
 #include <st/g4/stm32g474r(b-c-e)tx-pinctrl.dtsi>
+#include <zephyr/dt-bindings/adc/adc.h>
```
Please move this to "dts/am/st/g0/stm32g0.dtsi"

sambhurst replied:
Done

erwango commented on boards/arm/stm32g081b_eval/stm32g081b_eval.dts:
```diff
@@ -7,6 +7,7 @@
 /dts-v1/;
 #include <st/g0/stm32g081Xb.dtsi>
 #include <st/g0/stm32g081rbtx-pinctrl.dtsi>
+#include <zephyr/dt-bindings/adc/adc.h>
```
ditto

sambhurst replied:
Done

sambhurst committed:
dts: arm: st: include <dt-bindings/adc/adc.h>

Provide macros used for ADC channel configuration from a
devicetree node.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@erwango could you take another look when you have time? Thanks

erwango commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -1009,46 +1009,47 @@ static void ucpd_isr(const struct device *dev_inst[])
 				UCPD_SR_HRSTDISC;
 
-	if (IS_ENABLED(CONFIG_SOC_SERIES_STM32G0X)) {
+#ifdef CONFIG_SOC_SERIES_STM32G0X
```
Just a comment: What about checking `DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)` here ?
This could save some bytes in case only one instance is enabled, even on targets that support 2.

sambhurst replied:
I agree. I've made the change.

erwango commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -1009,46 +1009,47 @@ static void ucpd_isr(const struct device *dev_inst[])
 				UCPD_SR_HRSTDISC;
 
-	if (IS_ENABLED(CONFIG_SOC_SERIES_STM32G0X)) {
+#ifdef CONFIG_SOC_SERIES_STM32G0X
```
Nit: this doesn't match commit description

sambhurst replied:
The comment has been updated.

erwango commented on subsys/usb/usbc/Kconfig:
```diff
@@ -0,0 +1,44 @@
+
+menuconfig USBC_STACK
+	bool "USBC Stack Support"
+	default $(dt_compat_enabled,$(DT_COMPAT_USBC_PORT))
```
default y
depends on DT_HAS_USBC_PORT_ENABLED

sambhurst replied:
Done

erwango commented on samples/subsys/usbc/sink/prj.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_USBC_THREAD_PRIORITY=10
```
What about `CONFIG_USBC_TCPC_DRIVER`?

sambhurst replied:
This is enabled when the USBC stack is enabled.

menuconfig USBC_STACK
        bool "USBC Stack Support"
        default y
        depends on DT_HAS_USBC_PORT_ENABLED
        select SMF
        select SMF_ANCESTOR_SUPPORT
        select USBC_TCPC_DRIVER

erwango replied:
ok, right. But actually we should keep a high level Kconfig flag to allow user to enable the SW stack manually (just like we do for every subsystem or peripheral: ADC, SPI, USB_DEVICE_STACK, NETWORKING, ...).
So in this prj.conf, you should add `CONFIG_USBC_STACK=y`.
And `menuconfig USBC_STACK` should not use any `default y if foo` to avoid this level of control.

sambhurst replied:
I understand. Thanks

sambhurst committed:
stm32: Use DT_NUM_INST_STATUS_OKAY for num IRQ sources

Use DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) to determine
number of potential interrupt sources. This mitigates
the wasting of memory on devices with mutliple UCPD ports
but only one UCPD port is used.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

erwango commented on subsys/usb/usbc/Kconfig:
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2022 The Chromium OS Authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Workaround for not being able to have commas in macro arguments
+DT_COMPAT_USBC_PORT := usbc-port
```
Could be cleaned up

sambhurst replied:
I removed the lines.

erwango commented on subsys/usb/usbc/Kconfig:
```diff
@@ -0,0 +1,45 @@
+
+menuconfig USBC_STACK
+	bool "USBC Stack Support"
+	default y
+	depends on DT_HAS_USBC_PORT_ENABLED
```
See comment above. It should not be possible to enable this high level flag otherwise than manually.
Remove these lines.

sambhurst replied:
Done

erwango commented on drivers/usbc/tcpc/ucpd_stm32.c:
```diff
@@ -1009,46 +1009,58 @@ static void ucpd_isr(const struct device *dev_inst[])
+			break;
+		}
 	}
+#else
+#error "Multiple UCPD ports defined for device with only one port."
+#endif /* CONFIG_SOC_SERIES_STM32G0X */
```
This is not required. Device tree parsing will prevent this to happen.
The `#ifdef CONFIG_SOC_SERIES_STM32G0X` could be safely removed.

sambhurst replied:
Done

sambhurst commented:
@mbolivar-nordic could you take another look when you have time? Thanks

sambhurst committed:
boards: arm: stm32g081b_eval: Remove USB-C VBUS2 Discharge configuration

Remove USB-C VBUS2 Discharge configuration node from device tree.
USBC VBUS discharge is controlled by the vbus driver.
See dts/bindings/usb-c/vbus-adc.yaml

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@mbolivar-nordic could you take another look when you have time? Thanks

mbolivar-nordic reviewed (dismissed):
This is looking like it's in great shape now, but I don't see any documentation for the usbc_vbus_api group in Sphinx. Am I missing it? Please dismiss my review if so. Otherwise, please add it. No new APIs without at least some documentation, please (cc @nashif @carlescufi )

mbolivar-nordic reviewed (commented):
Please don't forget to update doc/develop/api/overview.rst as needed as well

sambhurst committed:
doc: add API documentation for USB-C VBUS Driver

This commit adds the USB-C driver API documentation in peripherals section
of the reference guide. The USB-C VBUS API is declared experimental.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@nashif  and @carlescufi this PR is adding a new subsystem and @mbolivar-nordic has requested that I CC you.

mbolivar-nordic commented:
> @nashif and @carlescufi this PR is adding a new subsystem and @mbolivar-nordic has requested that I CC you.

@sambhurst sorry if I was unclear -- when I wrote:

>  (cc @nashif @carlescufi )

that was actually **me** cc'ing them, not asking you to CC them

mbolivar-nordic commented:
>  I don't see any documentation for the usbc_vbus_api group in Sphinx. Am I missing it? Please dismiss my review if so. Otherwise, please add it.

@sambhurst are you going to address this comment? I see a new push since I asked this question, but it is still unaddressed.

sambhurst commented:
> > I don't see any documentation for the usbc_vbus_api group in Sphinx. Am I missing it? Please dismiss my review if so. Otherwise, please add it.
> 
> @sambhurst are you going to address this comment? I see a new push since I asked this question, but it is still unaddressed.

@mbolivar-nordic Yes I added the documentation in my last push.

galak reviewed (commented):
How does the bindings specified here related to the linux binding:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/connector/usb-connector.yaml

galak commented on drivers/usbc/tcpc/Kconfig.tcpc_stm32:
```diff
@@ -3,12 +3,10 @@
-
 config USBC_TCPC_STM32
 	bool "USBC TCPC device controller driver"
 	depends on SOC_FAMILY_STM32
```
remove the `depends on SOC_FAMILY_STM32`

sambhurst replied:
Done

galak commented on drivers/usbc/tcpc/Kconfig.tcpc_stm32:
```diff
@@ -3,12 +3,10 @@
 	bool "USBC TCPC device controller driver"
 	depends on SOC_FAMILY_STM32
-	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_UCPD))
+	default DT_HAS_ST_STM32_UCPD_ENABLED
```
make this 
```
default y
depends on DT_HAS_ST_STM32_UCPD_ENABLED
```

sambhurst replied:
Done

galak commented on drivers/usbc/vbus/Kconfig.vbus_adc:
```diff
@@ -0,0 +1,10 @@
+
+config VBUS_ADC
+	bool "VBUS ADC"
+	default DT_HAS_VBUS_ADC_ENABLED
```
Make this:
```
default y
depends on DT_HAS_VBUS_ADC_ENABLED
```

sambhurst replied:
Done

galak commented on dts/bindings/usbc/usbc-port.yaml:
```diff
@@ -0,0 +1,21 @@
+        type: phandle
+        required: true
+        description: |
+            VBUS measurement and control for this port.
```
what kinda of device is `vbus` pointing at?

sambhurst replied:
It's a device that can measure USBC VBUS and discharge it. Could be an ADC and a GPIO or a I2C TypeC Port Controller (TCPC). I've updated the description.

galak commented on dts/bindings/usbc/vbus-adc.yaml:
```diff
@@ -0,0 +1,13 @@
+
+description: USBC ADC Measurement and control
+
+compatible: "vbus-adc"
```
This compatible seems too generic.

sambhurst replied:
I changed the name to usbc-vbus-adc

galak commented on include/zephyr/drivers/usbc/usbc_vbus.h:
```diff
@@ -0,0 +1,114 @@
+extern "C" {
+#endif
+
+struct usbc_vbus_driver_api {
```
add __subsystem tag

sambhurst replied:
Done

galak reviewed (dismissed):
Some Kconfig changes needed, and questions about dts

sambhurst commented:
>

sambhurst commented:
@galak I was not aware of the Linux USB-C bindings but it seems I'm duplicating the Sink and VBUS portion.

sambhurst commented:
@galak and @mbolivar-nordic are there any other changes you'd like me to make. Thanks

galak commented:
> @galak I was not aware of the Linux USB-C bindings but it seems I'm duplicating the Sink and VBUS portion.

Can you move to using the linux bindings as this is one of our design goals

sambhurst committed:
dts: usb-c: Introduce usb-c-connector

A USB TypeC connector has many peripherals associated with
it and the DTS binding in this commit provides a way to group
peripherals and properties in a device tree.
This binding is used with the USB-C Subsytem.

This is based on Linux, documentation:
https://www.kernel.org/doc/Documentation/devicetree/bindings/connector/usb-connector.yaml

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
> > @galak I was not aware of the Linux USB-C bindings but it seems I'm duplicating the Sink and VBUS portion.
> 
> Can you move to using the linux bindings as this is one of our design goals

@galak I've included the linux bindings for the USB-C Connector.

sambhurst commented:
@galak Please take another look when you have time. Thanks

mbolivar-nordic reviewed (dismissed):
I didn't review this all in detail but it seems reasonable at this point from a high level. Thanks for all the updates!

galak reviewed (commented):
Is there any expectation of the USBC APIs working with userspace?

galak reviewed (commented):
Any plans to update MAINTAINERS.yml for this?

Who will be maintaining this code?

galak commented on include/zephyr/usbc/usbc.h:
```diff
@@ -0,0 +1,290 @@
+ *		No capability mismatch
+ *		Don't giveback
+ *		Object position 1 (5V PDO)
+ */
```
Feels like we need to namespace some of these enums and defines.  `policy_request_t` seems a bit too generic.

sambhurst replied:
I've prefixed the enums with "usbc_"

galak reviewed (commented):
@carlescufi should the PR update doc/develop/api/overview.rst?

sambhurst committed:
MAINTAINERS: Add USB-C maintainer

Add a new entry specifically for USB-C.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
> Is there any expectation of the USBC APIs working with userspace?

@galak Currently there are no expectations for the USBC APIs to work with userspace, although this feature could be added at a later date.

sambhurst commented:
> Any plans to update MAINTAINERS.yml for this?
> 
> Who will be maintaining this code?

@galak I have added myself as the maintainer of the code and I'm sure there will be many contributors.

sambhurst commented:
> @carlescufi should the PR update doc/develop/api/overview.rst?

@galak I've updated the overview.rst file.

galak commented:
> @galak Currently there are no expectations for the USBC APIs to work with userspace, although this feature could be added at a later date.

So my comment earlier about adding __subsystem was wrong, please remove that since there is no expectation for userspace support.

sambhurst commented:
> So my comment earlier about adding __subsystem was wrong, please remove that since there is no expectation for
> userspace support.

@galak I've removed __subsystem

galak commented on MAINTAINERS.yml:
```diff
@@ -2060,6 +2060,19 @@ USB:
   labels:
     - "area: USB"
 
+USBC:
```
nit pick: `USBC` -> `USB-C` (little more readable)

sambhurst replied:
done

galak commented on MAINTAINERS.yml:
```diff
@@ -2060,6 +2060,19 @@ USB:
+    - include/zephyr/usbc/
+    - samples/subsys/usbc/
+  labels:
+    - "area: USBC"
```
nit pick: `area: USBC` -> `area: USB-C` (little more readable)

sambhurst replied:
done

galak reviewed (commented):
General: think about places that using `usb-c` might be a little easier to read than `usbc` (dir names, file names, etc).

sambhurst committed:
usb-c: Change dir name from usbc to usb_c

For improved readability, change dir name from
usbc to usb_c

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
> General: think about places that using `usb-c` might be a little easier to read than `usbc` (dir names, file names, etc).

@galak I went through the PR and updated the names from usbc to usb_c

jfischer-no commented on MAINTAINERS.yml:
```diff
@@ -2060,6 +2060,21 @@ USB:
+    - dts/bindings/usb-c/
+    - include/zephyr/*/usb_c/
+    - include/zephyr/usb_c/
+    - samples/subsys/usb_c/
```
Please add
```
    - subsys/usb/usb_c/

```

possibly also to CODEOWNERS file.

sambhurst replied:
done

sambhurst committed:
CODEOWNERS: Add ownership for USB-C

Add myself as codeowner for the USB-C subsystem and sample

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@galak and @jfischer-no are there any other requested changed?

sambhurst commented:
@galak and @jfischer-no are there any other requested changed?

sambhurst commented:
@galak and @jfischer-no are there any other requested changed?

carlescufi commented:
CC @desowin

desowin commented on include/zephyr/drivers/usb_c/usbc_pd.h:
```diff
@@ -97,6 +97,104 @@ extern "C" {
+ * @brief Minimum time a response must be sent from a Port Partner
+ *        See Table 6-68 Time Values
+ */
+#define PD_T_SENDER_RESPONSE_MIN_MS 24
```
USB Power Delivery Specification Revision 3.1, Version 1.5, Table 6-68 Time Values, tSenderResponse contains values min 26, nom 29, max 32 ms.

Is there a difference between USB Power Delivery Specification Revision 3.0 Version 2.0 (as noted in the usbc_pd.h header) and USB Power Delivery Specification Revision 3.1, Version 1.5 (the version currently available for download at usb.org Document Library) or are the values entered incorrectly here?

sambhurst replied:
Yes. It seems tSenderResponse changed in Revision 3.1, Version 1.4.
**Rev 3.1, Version 1.3**: 24 min, 27 nom, 30 max
**Rev 3.1, Version 1.4**: 26 min, 29 nom, 32 max

desowin commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,214 @@
+    when it is attached to both MUX and USB-C interface controller.
+
+    This is based on Linux, documentation:
+      https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/connector/usb-connector.yaml
```
Linking to a repository that has rebasing in name is questionable. Also, the link is to latest version and not to the actual version that this was based on.

Did you base it on some rc version (possibly already rebased)? If you based it on Linux kernel v5.19, then 
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/connector/usb-connector.yaml?h=v5.19&id=3d7cb6b04c3f3115719235cc6866b10326de34cd
would be a better link

sambhurst replied:
I've updated the link. Thanks

desowin replied:
WHile you updated the code, the commit message still has the old link.

sambhurst replied:
Done

jfischer-no commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,214 @@
+          - "1.5A"
+          - "3.0A"
+        description: |
+          Determines the power operation mode that the Type C connector
```
Properties documentation is exactly the same as in the link above in commit message, it is problematic, compatibility is one thing, but the documentation should be written from scratch.

sambhurst replied:
I've updated the documentation.

jfischer-no reviewed (changes requested):
Properties documentation copyright issue should be fixed before this PR can go in.

desowin commented on include/zephyr/dt-bindings/usb-c/pd.h:
```diff
@@ -0,0 +1,94 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This is based on Linux, documentation:
+ *  https://github.com/torvalds/linux/blob/master/include/dt-bindings/usb/pd.h
```
Linking to master is really bad idea. You can use https://github.com/torvalds/linux/blob/v5.19/include/dt-bindings/usb/pd.h instead (if you based it on 5.19)

The github links are shorter than git.kernel.org so you might want to use the github link in usb-c-connector.yaml (and commit message) as well.

sambhurst replied:
Done

desowin commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -0,0 +1,290 @@
+	MSG_ACCEPT_RECEIVED,
+	/** Power Delivery Reject message was received */
+	MSG_REJECTED_RECEIVED,
+	/** Power Deleivery discarded the message being transmited */
```
Deleivery -> Delivery

sambhurst replied:
Done

desowin reviewed (commented):
Finally managed to skim through the whole PR. In general it looks ok, but my Zephyr specific knowledge is currently very limited. I think it is important to make the links to fixed revisions (instead of master branch) as otherwise the link will point to something different soon (if pointing to specific version, even if the server goes offline, itis still possible to deduct what version of the file was mentioned by looking at the link itself).

jfischer-no commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,173 @@
+    when it is attached to both MUX and USB-C interface controller.
+
+    This is based on Linux, documentation:
+      https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/connector/usb-connector.yaml?h=v5.19&id=3d7cb6b04c3f3115719235cc6866b10326de34cd
```
The link, also in commit message, should be https://www.kernel.org/doc/Documentation/devicetree/bindings/connector/usb-connector.yaml

jfischer-no replied:
git grep "based on Linux" dts/

sambhurst replied:
Done

galak commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig USBC_STACK
+	bool "USBC Stack Support"
```
nit: 'USB-C Stack Support`

sambhurst replied:
Done

galak commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	  Set thread priority of the USBC
+
+config USBC_STACK_SIZE
+	int "USBC thread stack size"
```
`USB-C`

sambhurst replied:
Done

galak commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	  Stack size of thread created for each instance.
+
+config USBC_STATE_MACHINE_CYCLE_TIME
+	int "USBC state machine cycle time in milliseconds"
```
`USB-C`

sambhurst replied:
Done

galak commented on dts/bindings/usb-c/usb-c-vbus-adc.yaml:
```diff
@@ -0,0 +1,15 @@
+    This device is used to measure VBUS on a Type-C Port and to
+    discharge VBUS when needed.
+
+compatible: "usb-c-vbus-adc"
```
should be `zephyr,usb-c-vbus-adc` (rename file as well).

sambhurst replied:
Done

galak commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,173 @@
+        type: string
+        required: true
+        enum:
+          - "SINK"
```
should be `lowercase`

sambhurst replied:
Done

galak commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,173 @@
+        type: string
+        required: false
+        enum:
+          - "SINK"
```
should be `lowercase`

sambhurst replied:
Done

galak commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,173 @@
+        type: string
+        required: false
+        enum:
+          - "UFP"
```
can these not match the linux binding of `host`, `device`, and `dual`?

sambhurst replied:
The spec and all the macros use UFP an DFP. I thought it would be clearer if we used UFP and DFP instead of Device and Host.

galak replied:
we should be trying to match the linux binding as much as possible.  So I think its better to use `host` and `device` and just add a comment in the binding about those mapping to `UFP` and `DFP`

sambhurst replied:
Done

galak commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -0,0 +1,173 @@
+
+    This is based on Linux, documentation:
+      https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/connector/usb-connector.yaml?h=v5.19&id=3d7cb6b04c3f3115719235cc6866b10326de34cd
+
```
Looks like we are missing the `ports` structure from the linux binding.  Any reason for this?

sambhurst replied:
The ports defined in Linux have a slightly different representation. For example, it defines ports as follows:
```
     port@0:
        description: High Speed (HS), present in all connectors.

      port@1:
        description: Super Speed (SS), present in SS capable connectors.

      port@2:
        description: Sideband Use (SBU), present in USB-C. This describes the
          alternate mode connection of which SBU is a part.
```
I thought this was to restrictive and decided to leave the port definition up to the user.

galak replied:
Does one of these not map to what you are trying to describe?

(as above, we should be trying to match the linux binding as much as possible.  If one of these port@N doesn't match what you need, than I think its reasonable to add a new port@3, port@4, etc..).  But we should match the structure if we can.

sambhurst replied:
The Linux port combine USB and USB-C. Currently, in zephyr, the two busses are separate. port@0 refers to USB 2.0 and port@1 refers to USB 3.0. The Sideband Use (SBU) on port@2 is also not used in the stack.

Combining USB and USB-C is probably something we should discuss in an Issue and implement in a separate PR.

galak replied:
What is the actual HW configuration we are trying to describe?

sambhurst replied:
We can only implement a USB-C Sink Power Delivery device and each port should have a Type-C Port Controller (TCPC) and a way to measure USB-C VBUS (Some TCPCs can be used to measure VBUS).

```
port1: usbc-port@1 {
                        compatible = "usb-c-connector";
                        reg = <1>;
                        tcpc = <&ucpd1>;
                        vbus = <&vbus1>;
                        power-role = "sink";
                        sink-pdos = <PDO_FIXED(5000, 100, 0)>;
                };
```

galak replied:
I don't know enough about usb to parse that :).

Put the actual software aside, is it possible to describe the HW we want to support using the linux binding or not?  If so, what would that dts (nodes) look like (based purely on the linux binding)?

sambhurst replied:
The short answer is No. The Linux ports are mostly concerned with USB data and not USB-C power delivery.

galak replied:
What's the disconnect between the two?  Seems like the binding is trying to describe both cases for a usb-connector.   Or are you saying that specifically with regards to `port@N`?

galak replied:
I'm glancing at some DTS files in linux and see things like:

```
        ptn5110: tcpc@50 {
                compatible = "nxp,ptn5110";
                pinctrl-names = "default";
                pinctrl-0 = <&pinctrl_typec1>;
                reg = <0x50>;
                interrupt-parent = <&gpio2>;
                interrupts = <11 IRQ_TYPE_LEVEL_LOW>;
                status = "okay";

                port {
                        typec1_dr_sw: endpoint {
                                remote-endpoint = <&usb1_drd_sw>;
                        };
                };

                typec1_con: connector {
                        compatible = "usb-c-connector";
                        label = "USB-C";
                        power-role = "dual";
                        data-role = "dual";
                        try-power-role = "sink";
                        source-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)>;
                        sink-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)
                                     PDO_VAR(5000, 20000, 3000)>;
                        op-sink-microwatt = <15000000>;
                        self-powered;
                };
        };
};
```

or
```
        wusb3801: tcpc@60 {
                compatible = "willsemi,wusb3801";
                reg = <0x60>;
                interrupt-parent = <&gpio0>;
                interrupts = <RK_PA4 IRQ_TYPE_LEVEL_LOW>;
                pinctrl-0 = <&tcpc_int_l>;
                pinctrl-names = "default";

                connector {
                        compatible = "usb-c-connector";
                        label = "USB-C";
                        vbus-supply = <&otg_switch>;
                        power-role = "dual";
                        try-power-role = "sink";
                        data-role = "dual";
                        typec-power-opmode = "default";
                        pd-disable;

                        ports {
                                #address-cells = <0x1>;
                                #size-cells = <0x0>;

                                port@0 {
                                        reg = <0x0>;

                                        typec_hs_usb2phy0: endpoint {
                                                remote-endpoint = <&usb2phy0_typec_hs>;
                                        };
                                };
                        };
                };
        };
```

sambhurst replied:
From the Linux ports description:
```
description: OF graph bindings modeling any data bus to the connector
      unless the bus is between parent node and the connector. Since a single
      connector can have multiple data buses every bus has an assigned OF graph
      port number as described below.
```
From the USB-C stacks port of view, it doesn't care about USB data bus or  Sideband SBU connections.

The USB-C stack in this PR only cares about Sink Power Delivery, not anything USB

sambhurst replied:
From the example you provided
```
port@0 {
                                        reg = <0x0>;

                                        typec_hs_usb2phy0: endpoint {
                                                remote-endpoint = <&usb2phy0_typec_hs>;
                                        };
                                };
```
This is USB not USB-C.

sambhurst replied:
I'll add the Linux ports property but it won't be used by the USB-C stack. Could I do the addition in a separate PR?

galak replied:
the devicetree is trying to describe hardware.  So lets ignore what the various software stacks do or don't do with that data.

In the examples I provided are `tcpc@50` and `tcpc@60` the type `tcpc` controller?

What I'm trying to focus on are the `tcpc` and `vbus` properties.  I think I understand how `port` isn't relevant at this time for you for what you need described.

galak replied:
@keith-zephyr if the diagram, is the gray pill box the 'usb-c-connector'?

keith-zephyr replied:
Yes, that is correct.

keith-zephyr replied:
The Linux USB Connector node is largely defining USB policy information.

Does it make sense for Zephyr to create a "usb-c-pd" node type to specify the chips (TCPC, PPC, MUXes), and to make the USB Connector a child?

```
	ports {
		#address-cells = <1>;
		#size-cells = <0>;
		/* usbc.rst usbc-port start */
		port1: usbc-port@1 {
			compatible = "usb-c-pd";
			reg = <1>;
			tcpc = <&ucpd1>;
			/* Note - ppc/muxes not part of this PR but needed later */
			ppc = <&ppc1>;
			muxes = <&retimer1 &mux1>;
			connector {
			    compatible = "usb-c-connector";
			    power-role = "dual";
			    vbus-supply = <&vbus1>;
			    ... /* Other properties: vdos, data-role, etc */ 
			};
		};
	};
```

galak replied:
Here are some links to the commits on the linux side that seem associated with google:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=d76a3c08d6496674da8dbb17e95546852f60e0ea
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=f28dc9ecd6697420b0777a99a5d1bc97f1a0d857
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=9af0f74778519a0f8bf22eddf5ad6c87b1223321

galak reviewed (dismissed):
need some changes to the bindings, and updates to places that we could use `usb-c`

sambhurst commented:
@galak ping

jfischer-no commented:
:hourglass_flowing_sand: :smile_cat:

galak reviewed (changes requested):
-1, until devicetree bindings are resolved.

mmahadevan108 commented:
@sambhurst I have deleted the 3.2 milestone as we are past the feature freeze date. Please pick a milestone from a future release.

sambhurst commented:
> > The Linux USB Connector node is largely defining USB policy information.
> > Does it make sense for Zephyr to create a "usb-c-pd" node type to specify the chips (TCPC, PPC, MUXes), and to make the USB Connector a child?
> > ```
> > 	ports {
> > 		#address-cells = <1>;
> > 		#size-cells = <0>;
> > 		/* usbc.rst usbc-port start */
> > 		port1: usbc-port@1 {
> > 			compatible = "usb-c-pd";
> > 			reg = <1>;
> > 			tcpc = <&ucpd1>;
> > 			/* Note - ppc/muxes not part of this PR but needed later */
> > 			ppc = <&ppc1>;
> > 			muxes = <&retimer1 &mux1>;
> > 			connector {
> > 			    compatible = "usb-c-connector";
> > 			    power-role = "dual";
> > 			    vbus-supply = <&vbus1>;
> > 			    ... /* Other properties: vdos, data-role, etc */ 
> > 			};
> > 		};
> > 	};
> > ```
> 
> Something like this seems like a good direction to me. How would we describe a system with `usb-c-pd` and usb "data" on the same usb-c port/connector?
> 
> Any ability to reach out to the google guys that worked on the linux binding?
> 
> what might be examples of PPC, are their chips that just do this functionality?

@galak I had a meeting with one of the Googler's, Badhri Sridharan <badhri@google.com>, that worked on the Linux bindings   and his group is basically using the usb-c-connector similarly to the example device tree above.

Also, this PR has grown relatively large. Do you think it would be okay to merge it as is and move the USB-C binding discussion to another PR? Or I could create an Issue for the USB-C bindings and have the discussion there.

keith-zephyr commented:
@galak Can you re-review?  Are you okay with landing this as is with a follow up PR to specifically address the devicetree?

galak reviewed (dismissed):
DTS LGTM for now, we can re-visit in the future as needed.

carlescufi commented:
@sambhurst this is almost ready, could you please rebase and solve the conflict?

sambhurst commented:
@carlescufi  The rebase is done and all conflict are resolved

sambhurst commented:
@galak and @erwango could you take another look after my rebase? Thanks

nashif commented on doc/connectivity/index.rst:
```diff
@@ -9,3 +9,4 @@ Connectivity
    bluetooth/index.rst
    networking/index.rst
    lora_lorawan/index.rst
+   usb_c/index.rst
```
does usb-c really belong under connectivity?

carlescufi replied:
From my perspective, yes. It's not wireless, but it's definitely connectivity. I think we should move the general USB doc to connectivity in fact. I will send a PR after this one is merged.

nashif replied:
connectivity here was meant to accomodate comm technologies, wired and wireless. For the sake of this PR going in, we can revisit this later and discusss what falls under connectivity somewhere else.
